rust_dev_preview/s3/src/bin/put-object-progress.rs

```diff
@@ -0,0 +1,190 @@
+use std::{
+    mem,
+    path::PathBuf,
+    pin::Pin,
+    process::exit,
+    task::{Context, Poll},
+};
+
+use aws_config::meta::region::RegionProviderChain;
+use aws_sdk_s3::{
+    primitives::{ByteStream, SdkBody},
+    Client,
+};
+use aws_smithy_http::body::BoxBody;
+use bytes::Bytes;
+use clap::Parser;
+use http_body::{Body, SizeHint};
+use tracing::{debug, info};
+
+#[derive(Debug, Parser)]
+struct Opt {
+    #[structopt(long)]
+    bucket: String,
+    #[structopt(long)]
+    object: String,
+    #[structopt(long)]
+    source: PathBuf,
+}
+
+// ProgressTracker prints information as the upload progresses.
+struct ProgressTracker {
+    bytes_written: u64,
+    content_length: u64,
+}
+
+impl ProgressTracker {
+    fn track(self: &mut Self, len: u64) -> () {
+        self.bytes_written += len;
+        let progress = self.bytes_written as f64 / self.content_length as f64;
+        info!("Read {} bytes, progress: {:.2}%", len, progress * 100.0);
+    }
+}
+
+// snippet-start:[s3.rust.put-object-progress-body]
+// A ProgressBody to wrap any http::Body with upload progress information.
+#[pin_project::pin_project]
+pub struct ProgressBody<InnerBody> {
+    #[pin]
+    inner: InnerBody,
+    // prograss_tracker is a separate field so it can be accessed as &mut.
+    progress_tracker: ProgressTracker,
+}
+
+// For an SdkBody specifically, the ProgressTracker swap itself in-place while customizing the SDK operation.
+impl ProgressBody<SdkBody> {
+    // Wrap a Requests's SdkBody with a new ProgressBody, and replace it on the fly.
+    // This is specialized for SdkBody specifically, as SdkBody provides ::taken() to
+    // swap out the current body for a fresh, empty body and then provides ::from_dyn()
+    // to get an SdkBody back from the ProgressBody it created. http::Body does not have
+    // this "change the wheels on the fly" utility.
+    pub fn replace(
+        mut value: http::Request<SdkBody>,
+    ) -> Result<http::Request<SdkBody>, anyhow::Error> {
+        let body = mem::replace(value.body_mut(), SdkBody::taken()).map(|body| {
+            let len = body.content_length().expect("upload body sized"); // TODO - panics
```

Comment:
this [.map](https://github.com/awslabs/smithy-rs/blob/7cb9d6af3ea4c0f55c65c662f1e7b62880bc7728/rust-runtime/aws-smithy-http/src/body.rs#L187) _must_ return an SdkBody, so it can't try_into or similar here. (I forgot I put the note in, glad I did and that you caught it). I'm not sure if SdkBody is at fault here or if this code is - maybe ProgressBody should be smarter to handle None sized bodys?